Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
+{
+    // Traverse all children
+    for (auto* child : m_children) {
+        // Recursively collect from the child's descendants
+        child->collectLayersToFlattenRecursive(layers);
+
+        // If the child satisfies the condition, add it to the layers vector
+        if (child->flattensAsLeafOf3DSceneOr3DPerspective())
+            layers.append(child);
+    }
+}
+
+void TextureMapperLayer::flatten(TextureMapperPaintOptions& options)
+{
+    // Reset transformations for flattening root
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
+    SetForScope scopedFutureLocalTransform(m_layerTransforms.futureLocalTransform, TransformationMatrix());
+    m_layerTransforms.combined = TransformationMatrix();
+    m_layerTransforms.combinedForChildren = TransformationMatrix();
+    m_layerTransforms.futureLocalTransform = TransformationMatrix();
+    m_layerTransforms.futureCombined = TransformationMatrix();
+    m_layerTransforms.futureCombinedForChildren = TransformationMatrix();
+    ComputeTransformData data;
+    if (m_state.maskLayer)
+        m_state.maskLayer->computeTransformsRecursive(data);
+    if (m_state.replicaLayer)
+        m_state.replicaLayer->computeTransformsRecursive(data);
+    if (m_state.backdropLayer)
+        m_state.backdropLayer->computeTransformsRecursive(data);
+    for (auto* child : m_children)
+        child->computeTransformsRecursive(data);
+
+    // Flatten the layer
+    if (m_flattenedLayer != nullptr) {
+        delete m_flattenedLayer;
+        m_flattenedLayer = nullptr;
+    }
+
+    Region layerRegion;
+    computeFlattenedRegion(layerRegion);
+
+    auto flattenedLayerRect = layerRegion.bounds();
+    auto flattenedLayer = new TextureMapperFlattenedLayer(flattenedLayerRect);
+
+    flattenedLayer->bind(options.textureMapper);
+    {
+        SetForScope scopedSurface(options.surface, flattenedLayer->surface());
+        SetForScope scopedOffset(options.offset, -IntSize(flattenedLayerRect.x(), flattenedLayerRect.y()));
+        SetForScope scopedOpacity(options.opacity, 1);
+
+        paintSelfAndChildren(options, false);
+    }
+
+    m_flattenedLayer = flattenedLayer;
+    options.textureMapper.bindSurface(options.surface.get());
+}
+
+void TextureMapperLayer::computeFlattenedRegion(Region& region)
+{
+    FloatRect rect = layerRect();
+
+    bool shouldExpand = m_currentFilters.hasOutsets() && !m_state.masksToBounds && !m_state.maskLayer;
+    if (shouldExpand) {
+        auto outsets = m_currentFilters.outsets();
+        rect.move(-outsets.left(), -outsets.top());
+        rect.expand(outsets.left() + outsets.right(), outsets.top() + outsets.bottom());
+    }
+
+    region.unite(enclosingIntRect(m_layerTransforms.combined.mapRect(rect)));
+
+    if (!m_state.masksToBounds && !m_state.maskLayer) {
+        for (auto* child : m_children)
+            child->computeFlattenedRegion(region);
+    }
+}
+
+void TextureMapperLayer::postprocess()
```

Comment:
Why divide this in postprocess / freeFlattenedLayersRecursive?

Reply:
Fixed